routers/web/events/websocket.go

```diff
@@ -0,0 +1,235 @@
+			}
+
+			return u.Host == appURLURL.Host
+		},
+	}
+
+	// Because http proxies will tend not to pass these headers
+	ctx.Req.Header.Add("Upgrade", "websocket")
+	ctx.Req.Header.Add("Connection", "Upgrade")
+
+	conn, err := upgrader.Upgrade(ctx.Resp, ctx.Req, nil)
+	if err != nil {
+		log.Error("Unable to upgrade due to error: %v", err)
+		return
+	}
+	defer conn.Close()
+
+	notify := ctx.Done()
+	shutdownCtx := graceful.GetManager().ShutdownContext()
+
+	eventChan := make(<-chan *eventsource.Event)
+	uid := int64(0)
+	unregister := func() {}
+	if ctx.IsSigned {
+		uid = ctx.Doer.ID
+		eventChan = eventsource.GetManager().Register(uid)
+		unregister = func() {
+			go func() {
+				eventsource.GetManager().Unregister(uid, eventChan)
+				// ensure the messageChan is closed
+				for {
+					_, ok := <-eventChan
+					if !ok {
+						break
+					}
+				}
+			}()
+		}
+	}
+	defer unregister()
+
+	readMessageChan := make(chan readMessage, readMessageChanSize)
+	go readMessagesFromConnToChan(conn, readMessageChan)
+
+	pingTicker := time.NewTicker(pingPeriod)
+
+	for {
+		select {
+		case <-notify:
+			return
+		case <-shutdownCtx.Done():
+			return
+		case <-pingTicker.C:
+			// ensure that we're not already cancelled
+			select {
+			case <-notify:
+				return
+			case <-shutdownCtx.Done():
+				return
+			default:
+			}
+
+			if err := conn.SetWriteDeadline(time.Now().Add(writeWait)); err != nil {
+				log.Error("unable to SetWriteDeadline: %v", err)
+				return
+			}
+			if err := conn.WriteMessage(websocket.PingMessage, []byte{}); err != nil {
+				log.Error("unable to send PingMessage: %v", err)
+				return
+			}
+		case message, ok := <-readMessageChan:
+			if !ok {
+				break
+			}
+
+			// ensure that we're not already cancelled
+			select {
+			case <-notify:
+				return
+			case <-shutdownCtx.Done():
+				return
+			default:
+			}
+
+			// FIXME: HANDLE MESSAGES
+			log.Info("Got Message: %d:%s:%v", message.messageType, message.message, message.err)
+		case event, ok := <-eventChan:
+			if !ok {
+				break
+			}
+
+			// ensure that we're not already cancelled
+			select {
+			case <-notify:
+				return
+			case <-shutdownCtx.Done():
+				return
+			default:
+			}
+
+			// Handle events
+			if event.Name == "logout" {
+				if ctx.Session.ID() == event.Data {
+					event = &eventsource.Event{
+						Name: "logout",
+						Data: "here",
+					}
+					_ = writeEvent(conn, event)
+					go unregister()
+					auth.HandleSignOut(ctx)
+					break
+				}
+				// Replace the event - we don't want to expose the session ID to the user
+				event = &eventsource.Event{
+					Name: "logout",
+					Data: "elsewhere",
+				}
+			}
+			if err := writeEvent(conn, event); err != nil {
+				return
+			}
+		}
+	}
+}
+
+func readMessagesFromConnToChan(conn *websocket.Conn, messageChan chan readMessage) {
+	defer func() {
+		close(messageChan) // Please note: this has to be within a wrapping anonymous func otherwise it will be evaluated when creating the defer
```

Comment:
Interesting, I didn't know this. Is this something specific to the `close` builtin?